[CPU] tails optimization for load/store emitters

chenhu-wang opened the pull request:
### Details:
 - *update tails load/store emulation method* 
 - *introduce smart number between mask load/store and emulation on avx512 platform*
 - *remove unnecessary instructions*

### Tickets:
 - *87986*

dmitry-gorokhov commented:
@a-sidorova please review the PR

a-sidorova commented on src/plugins/intel_cpu/src/emitters/jit_load_store_emitters.cpp:
```diff
@@ -185,7 +185,10 @@ void jit_load_emitter::load_bytes(const Vmm &vmm, const Xbyak::Reg64 &reg, int o
             case 0: break;
-            case 1: h->uni_vpinsrb(xmm, xmm, addr(start_bytes), 0); break;
+            case 1: // h->uni_vpinsrb(xmm, xmm, addr(start_bytes), 0); break;
+                h->movzx(Reg32(aux_gpr_idxs[0]), addr(start_bytes));
+                h->uni_vmovq(xmm, Reg64(aux_gpr_idxs[0]));
```
Have you checked the same changes for store emitter?
In [Interpolate](https://github.com/openvinotoolkit/openvino/blob/releases/2022/1/src/plugins/intel_cpu/src/nodes/interpolate.cpp#L1443-L1444) there are the same lines for store emitter.

chenhu-wang replied:
The interesting thing is, for store emitter, converting to traditional store_scalar() manner does not have perf improvement, at least I tested on ADL and TGL. Anyway, I replaced with the mov_x() for store scalar, as show no regression as well, in case there are regression on some platforms or code context I have not cover. Thanks!

dmitry-gorokhov commented:
In general LGTM. Would be nice if we can run internal perf validation before merge.

chenhu-wang commented:
> In general LGTM. Would be nice if we can run internal perf validation before merge.

Sure, internal perf validation is WIP, thanks

dmitry-gorokhov commented:
@chenhu-wang  do we have perf validation results?

chenhu-wang commented:
> @chenhu-wang do we have perf validation results?

@dmitry-gorokhov , the result is updated in ticket. Thanks!

dmitry-gorokhov commented:
@chenhu-wang Please rebase the branch

chenhu-wang commented:
> @chenhu-wang Please rebase the branch

@dmitry-gorokhov, I have rebased on latest master, there are some failure appeared for convert and split node on CI on Windows. WIP to analysis. @a-sidorova appreciate comment if you have some insights. aux_gpr_reg maybe handled incorrectly I guess.

chenhu-wang commented:
> 

@a-sidorova , Thanks a lot for your investigation and input. For rsp, rbp, rsi, rdi, it's lower 8bit part reg does not have the same index with 64 bit reg. Because there are more than 16 8bit regs. e.g. rax has al and ah. The correct idx is 32+index for these 4 gpr.   Correct it in [63f1063](https://github.com/openvinotoolkit/openvino/pull/12363/commits/63f106317b131c51947e5bd85744b9ab807199db).

chenhu-wang commented:
> > @chenhu-wang Please rebase the branch
> 
> @dmitry-gorokhov, I have rebased on latest master, there are some failure appeared for convert and split node on CI on Windows. WIP to analysis. @a-sidorova appreciate comment if you have some insights. aux_gpr_reg maybe handled incorrectly I guess.

@dmitry-gorokhov ,rebased and CI get green, thank you!

chenhu-wang commented:
> > 
> 
> @a-sidorova , Thanks a lot for your investigation and input. For rsp, rbp, rsi, rdi, it's lower 8bit part reg does not have the same index with 64 bit reg. Because there are more than 16 8bit regs. e.g. rax has al and ah. The correct idx is 32+index for these 4 gpr. Correct it in [63f1063](https://github.com/openvinotoolkit/openvino/pull/12363/commits/63f106317b131c51947e5bd85744b9ab807199db).

@a-sidorova @dmitry-gorokhov , the conversion happens [here](https://github.com/openvinotoolkit/oneDNN/blob/10c74ec7536e7b13cce013f9a229c217d7ace7ed/src/cpu/x64/xbyak/xbyak.h#L521).

dmitry-gorokhov commented on src/plugins/intel_cpu/src/emitters/jit_load_store_emitters.cpp:
```diff
@@ -751,9 +755,17 @@ void jit_store_emitter::store_bytes(const Vmm &vmm, const Xbyak::Reg64 &reg, int
+                    ext8bit = true;
+                h->mov(addr(start_bytes), Reg8(aux_gpr_idxs[0], ext8bit));
+                break;
+                // h->uni_vpextrb(addr(start_bytes), xmm, 0); break;
```
To be honest the comment is a little bit confusing. Lets either delete it or add a note that this instruction is replaced with pure mov by performance reasons.

chenhu-wang replied:
removed, thanks!

dmitry-gorokhov commented on src/plugins/intel_cpu/src/emitters/jit_load_store_emitters.cpp:
```diff
@@ -751,9 +755,17 @@ void jit_store_emitter::store_bytes(const Vmm &vmm, const Xbyak::Reg64 &reg, int
             case 0: break;
-            case 1: h->uni_vpextrb(addr(start_bytes), xmm, 0); break;
+            case 1:
+                h->uni_vmovq(Reg64(aux_gpr_idxs[0]), xmm);
```
Does it really bring any perf improvement comparing with `uni_vpextrb`? If yes it looks weird why other cases (bytes: 2, 3, etc) are not changed as well.
This question applicable to load emitter as well,

chenhu-wang replied:
Based on experiment, there are perf improvement on small number(1-4 byte). Extend to these cases. Thank you Dmitry.

a-sidorova reviewed (changes requested):
As we discussed offline @chenhu-wang could you please cover case when `src_prc != dst_prc` as well?

It blocks my [PR#12395](https://github.com/openvinotoolkit/openvino/pull/12395). `load_emitter` and `store_emitter` are often used in the PR when `FakeQuantize` are inside `Subgrah` body. So we have performance degradations for `ScalarTile` in snippets when we want to load and store I8/U8 data

chenhu-wang commented:
> As we discussed offline @chenhu-wang could you please cover case when `src_prc != dst_prc` as well?
> 
> It blocks my [PR#12395](https://github.com/openvinotoolkit/openvino/pull/12395). `load_emitter` and `store_emitter` are often used in the PR when `FakeQuantize` are inside `Subgrah` body. So we have performance degradations for `ScalarTile` in snippets when we want to load and store I8/U8 data

This is extended, could you please taka a look? thank you.

a-sidorova commented on src/plugins/intel_cpu/src/emitters/jit_load_store_emitters.cpp:
```diff
@@ -745,22 +758,35 @@ void jit_store_emitter::store_bytes(const Vmm &vmm, const Xbyak::Reg64 &reg, int
         // tail 7 bytes for lower or upper xmm
+        auto store_one_byte = [&](int bytes_offset, int gpr_idx) {
+            bool ext8bit = false;
+            if (gpr_idx == Operand::RSP || gpr_idx == Operand::RBP ||
+                gpr_idx == Operand::RSI || gpr_idx == Operand::RDI)
```
```suggestion
            if (one_of(gpr_idx, Operand::RSP, Operand::RBP, Operand::RSI, Operand::RDI))
```

chenhu-wang replied:
done, thank you.

a-sidorova reviewed (commented):
Nice work! @chenhu-wang thank you very much for the covering `src != dst` case! Could you take a look at the rest of the comments, please?

chenhu-wang commented on src/plugins/intel_cpu/src/emitters/jit_load_store_emitters.cpp:
```diff
@@ -993,21 +1034,37 @@ void jit_store_emitter::store_dword_to_word_extension(const Vmm &vmm, const Xbya
+                } else {
+                    Vmm zero(aux_vec_idxs[0]);
+                    h->uni_vpxor(zero, zero, zero);
+                    h->uni_vpmaxsd(vmm, zero, vmm);        // if singed bit is 1, set value as 0.
```
exactly, (xmm, xmm, zero) could save one move instruction. Changed, thank you.

chenhu-wang commented:
@dmitry-gorokhov , could you please final review? Thanks.

dmitry-gorokhov reviewed (approved):
Please fix remaining comment before merge

chenhu-wang commented on src/plugins/intel_cpu/src/emitters/jit_load_store_emitters.cpp:
```diff
@@ -18,6 +18,11 @@ using namespace Xbyak::util;
 namespace intel_cpu {
 
+// heuristic threshold number by byte between mask load and emulation with several simple partial load
+const int threhold_for_mask_emu_load = 14;
```
oh, my bad, typo fixed in [a29969c](https://github.com/openvinotoolkit/openvino/pull/12363/commits/a29969c797245cf33e88555d383df7ab10affd5d). Thank you.
